File: stgit/commands/rebase.py
import itertools
import re
import typing
from enum import Enum

from stgit import utils
from stgit.argparse import opt
from stgit.commands.common import (
    CmdException,
    DirectoryGotoTopLevel,
    delete_patches,
    git_commit,
    post_rebase,
    prepare_rebase,
    rebase,
)
from stgit.commands.squash import squash
from stgit.config import config
from stgit.lib import edit
from stgit.run import RunException
from stgit.utils import edit_string

__copyright__ = """
Copyright (C) 2005, Catalin Marinas <catalin.marinas@gmail.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see http://www.gnu.org/licenses/.
"""

help = 'Move the stack base to another point in history'
kind = 'stack'
usage = ['[options] [--] [new-base-id]']
description = """
Pop all patches from current stack, move the stack base to the given
[new-base-id] and push the patches back.

If you experience merge conflicts, resolve the problem and continue
the rebase by executing the following sequence:

        $ git add --update
        $ stg refresh
        $ stg goto top-patch

Or if you want to skip that patch:

        $ stg undo --hard
        $ stg push next-patch..top-patch"""

args = ['commit']
options = [
    opt(
        '-i',
        '--interactive',
        action='store_true',
        short='Open an interactive editor to manipulate patches',
    ),
    opt(
        '-n',
        '--nopush',
        action='store_true',
        short='Do not push the patches back after rebasing',
    ),
    opt(
        '-m',
        '--merged',
        action='store_true',
        short='Check for patches merged upstream',
    ),
    opt(
        '--autostash',
        action='store_true',
        short='Stash changes before the rebase and apply them after',
        default=config.getbool('stgit.autostash'),
        long='''
    Automatically create a temporary stash before the operation begins, and
    apply it after the operation ends. This means that you can run rebase on a
    dirty work-tree. However, use with care: the final stash application after a
    successful rebase might result in non-trivial conflicts.
    ''',
    ),
]

directory = DirectoryGotoTopLevel()

INTERACTIVE_APPLY_LINE = '# --- APPLY_LINE ---'

INTERACTIVE_HELP_INSTRUCTIONS = """
# Commands:
# k, keep <patch> = do not modify this patch
# e, edit <patch> = interactively edit this patch
# s, squash <patch> = squash patch into the previous patch
# f, fixup <patch> = like "squash", but discard this patch's commit message
# d, delete <patch> = delete patch
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# Patches above the APPLY_LINE are applied; other patches are kept unapplied.
"""


class StashPopConflictException(RunException):
    """Exception raised there's a conflict when popping the stash after --autostash"""

    def __init__(self):
        super().__init__(
            'Merge conflict raised when popping stash after rebase. Resolve the '
            'conflict, then delete the stash with `git stash drop`.'
        )


def __get_description(stack, patch):
    """Extract and return a patch's short description"""
    cd = stack.patches[patch].data
    return cd.message_str.strip().split('\n', 1)[0].rstrip()


class Action(Enum):
    # do nothing; a 'no-op'
    KEEP = 1
    # edit the patch
    EDIT = 2
    # squash the patch into the previous patch
    SQUASH = 3
    # fix the patch into the previous patch
    FIXUP = 4
    # delete the patch
    DELETE = 5


Instruction = typing.NamedTuple(
    'Instruction', [('patch_name', str), ('action', Action), ('apply', bool)]
)


def __perform_squashes(instructions, index, stack):
    """Perform any consecutive squashes from 'index' onwards."""
    squashes_found = False
    # loop to determine how many of the next N instructions are squashes
    for num_squashes in itertools.count(1):  # pragma: no branch
        if len(instructions) <= index + num_squashes:
            break  # reached the end of the instruction list
        if instructions[index + num_squashes].action == Action.SQUASH:
            squashes_found = True
        else:
            break  # not a squash; chain is over
    if squashes_found:
        _, new_patch_name = squash(
            stack,
            stack.repository.default_iw,
            None,
            None,
            None,
            [p.patch_name for p in instructions[index : index + num_squashes]],
        )
        # update instruction set in case the current patch was renamed
        instructions[index] = Instruction(
            new_patch_name,
            instructions[index].action,
            instructions[index].apply,
        )
        # remove the squashed patches from the instruction set
        instructions = instructions[: index + 1] + instructions[index + num_squashes :]
    return instructions


def __fixup_patches(stack, iw, patches):
    assert len(patches) > 0
    base_patch = patches[0]
    cd = stack.patches[base_patch]
    squash(
        stack, iw, base_patch, cd.data.message_str, save_template=False, patches=patches
    )


def __do_rebase_interactive(repository, previously_applied_patches, check_merged):
    """Opens an interactive editor, generates instruction list, and executes instructions."""
    stack = repository.get_stack()

    if len(stack.patchorder.all) == 0:
        return utils.STGIT_SUCCESS

    name_len = max((len(s) for s in stack.patchorder.all))
    line = 'keep {{:<{name_len}}} # {{}}'.format(name_len=name_len)

    # create a list of all patches to send to the editor
    raw_editor_instructions = []
    for pn in previously_applied_patches:
        patch = (pn, __get_description(stack, pn))
        raw_editor_instructions.append(line.format(*patch))
    raw_editor_instructions.append(INTERACTIVE_APPLY_LINE)
    for pn in stack.patchorder.all:
        if pn in previously_applied_patches:
            continue
        patch = (pn, __get_description(stack, pn))
        raw_editor_instructions.append(line.format(*patch))
    raw_editor_instructions.append(INTERACTIVE_HELP_INSTRUCTIONS)

    # open an editor to let the user generate the 'todo' instructions
    todo = edit_string(
        '\n'.join(raw_editor_instructions),
        '.stgit-rebase-interactive.txt',
    )

    # parse the instructions we've been given
    seen_apply_line = False
    instructions = []
    for line in todo.splitlines():
        line = line.strip()

        # record when we find the APPLY_LINE so we know which patches to apply
        if INTERACTIVE_APPLY_LINE in line:
            if INTERACTIVE_APPLY_LINE == line:
                seen_apply_line = True
            else:
                raise CmdException("Bad APPLY_LINE: '%s'" % line)

        # ignore comment lines
        if not line or line.startswith('#'):
            continue

        # parse a single instruction
        match = re.match(r'(\S+) (\S+).*', line)
        if not match:
            raise CmdException("Bad todo line: '%s'" % line)
        instruction_str, patch_name = match.groups()
        if patch_name not in stack.patchorder.all:
            raise CmdException("Bad patch name '%s'" % patch_name)
        if instruction_str in ('k', 'keep'):
            instruction_type = Action.KEEP
        elif instruction_str in ('e', 'edit'):
            instruction_type = Action.EDIT
        elif instruction_str in ('s', 'squash'):
            instruction_type = Action.SQUASH
        elif instruction_str in ('f', 'fix', 'fixup'):
            instruction_type = Action.FIXUP
        elif instruction_str in ('d', 'delete'):
            instruction_type = Action.DELETE
        else:
            raise CmdException("Unknown instruction '%s'" % instruction_str)

        # save the instruction to execute later
        instructions.append(
            Instruction(patch_name, instruction_type, not seen_apply_line)
        )

    for index in itertools.count():  # pragma: no branch
        if index >= len(instructions):
            break  # reached the end of the instruction list
        patch_name = instructions[index].patch_name
        if instructions[index].action == Action.DELETE:
            delete_patches(stack, stack.repository.default_iw, {patch_name})
            index -= 1  # re-run this index another time
            continue
        if instructions[index].action == Action.FIXUP and index > 0:
            base_patch = instructions[index - 1].patch_name
            __fixup_patches(
                stack, stack.repository.default_iw, [base_patch, patch_name]
            )
            instructions.pop(index)  # remove 'fixed' (ie deleted) patch
            index -= 1  # re-run this index another time
            continue
        if instructions[index].action == Action.EDIT:
            cd = stack.patches[patch_name].data
            cd, new_patch_name, replacement_diff = edit.interactive_edit_patch(
                stack.repository, cd, patch_name, edit_diff=True, diff_flags=""
            )
            _, new_patch_name = edit.perform_edit(
                stack,
                cd,
                patch_name,
                new_patch_name,
                edit_diff=True,
                diff_flags="",
                replacement_diff=replacement_diff,
            )
            if new_patch_name:
                instructions[index] = Instruction(
                    new_patch_name,
                    instructions[index].action,
                    instructions[index].apply,
                )
                patch_name = new_patch_name
        if instructions[index].apply:
            post_rebase(stack, {patch_name}, 'rebase', check_merged)

        instructions = __perform_squashes(instructions, index, stack)

    return utils.STGIT_SUCCESS


def func(parser, options, args):
    """Rebase the current stack"""

    if options.interactive:
        # destination is optional for '--interactive'
        if len(args) not in (0, 1):
            parser.error('incorrect number of arguments')
    else:
        if len(args) != 1:
            parser.error('incorrect number of arguments')

    repository = directory.repository
    stack = repository.get_stack()
    iw = repository.default_iw

    if stack.protected:
        raise CmdException('This branch is protected. Rebase is not permitted')

    if len(args) > 0:
        target = git_commit(args[0], repository)
    else:
        target = stack.base

    if options.autostash and not iw.worktree_clean():
        repository.run(['git', 'stash', 'push']).run()
        stashed_worktree = True
    else:
        stashed_worktree = False

    applied = stack.patchorder.applied
    retval = prepare_rebase(stack, 'rebase')
    if retval:
        return retval
    rebase(stack, iw, target)

    if options.interactive:
        retval = __do_rebase_interactive(
            repository, applied, check_merged=options.merged
        )
    elif not options.nopush:
        retval = post_rebase(stack, applied, 'rebase', check_merged=options.merged)
    else:
        retval = None

    if stashed_worktree:
        try:
            repository.run(['git', 'stash', 'pop']).run()
        except RunException:
            raise StashPopConflictException()

    return retval

File: src/cmd/rebase.rs
// SPDX-License-Identifier: GPL-2.0-only

//! `stg rebase` implementation.

use std::{fmt::Write, rc::Rc, str::FromStr};

use anyhow::{anyhow, Result};
use bstr::ByteSlice;
use clap::{Arg, ArgMatches};

use crate::{
    argset,
    color::get_color_stdout,
    ext::RepositoryExtended,
    patch::{patchedit, PatchName},
    print_info_message,
    revspec::parse_stgit_revision,
    stack::{InitializationPolicy, Stack, StackAccess, StackStateAccess},
    stupid::Stupid,
};

pub(super) const STGIT_COMMAND: super::StGitCommand = super::StGitCommand {
    name: "rebase",
    category: super::CommandCategory::StackManipulation,
    make,
    run,
};

fn make() -> clap::Command {
    clap::Command::new(STGIT_COMMAND.name)
        .about("Move the stack base to another point in history")
        .long_about(
            "Pop all patches from the current stack, move the stack base to the given \
            new base and push the patches back.\n\
            \n\
            Merge conflicts may arise when patches are being pushed-back onto the \
            stack. If this occurs, resolve the conflicts and then continue the rebase \
            with the following sequence:\n\
            \n    \
            stg add --update\n    \
            stg refresh\n    \
            stg goto top-patch\n\
            \n\
            Or to skip the conflicting patch:\n\
            \n    \
            stg undo --hard\n    \
            stg push next-patch..top-patch\n\
            ",
        )
        .arg(
            Arg::new("committish")
                .help("New base commit for the stack")
                .value_parser(clap::builder::NonEmptyStringValueParser::new())
                .required_unless_present("interactive"),
        )
        .arg(
            Arg::new("interactive")
                .long("interactive")
                .short('i')
                .help("Interactively manipulate patches in editor")
                .action(clap::ArgAction::SetTrue),
        )
        .arg(
            Arg::new("nopush")
                .long("nopush")
                .short('n')
                .help("Do not push back patches after rebasing")
                .action(clap::ArgAction::SetTrue)
                .conflicts_with("merged"),
        )
        .arg(argset::merged_arg().long_help(
            "Check for patches that may have been merged upstream.\n\
             \n\
             When pushing-back patches, each patch is checked to see if its changes \
             already exist in the new stack base. If a patch's changes are detected to \
             have been merged, the patch will still exist in the stack, but become \
             empty after the rebase operation.",
        ))
        .arg(argset::committer_date_is_author_date_arg())
        .arg(
            Arg::new("autostash")
                .long("autostash")
                .help("Stash changes before the rebase and reapply them after")
                .long_help(
                    "Automatically create a temporary stash before the operation \
                     begins, and apply it after the operation completes. This allows a \
                     rebase to be performed on a dirty work tree. Note however that \
                     the final stash application may result in non-trivial conflicts.",
                )
                .action(clap::ArgAction::SetTrue),
        )
        .arg(argset::push_conflicts_arg())
}

fn run(matches: &ArgMatches) -> Result<()> {
    let repo = gix::Repository::open()?;
    let stack = Stack::from_branch(&repo, None, InitializationPolicy::RequireInitialized)?;
    let config = repo.config_snapshot();
    let stupid = repo.stupid();
    let branch_name = stack.get_branch_name().to_string();
    let allow_push_conflicts = argset::resolve_allow_push_conflicts(&config, matches);
    let committer_date_is_author_date = matches.get_flag("committer-date-is-author-date");

    let target_commit = if let Some(committish) = argset::get_one_str(matches, "committish") {
        Rc::new(
            parse_stgit_revision(&repo, Some(committish), None)?
                .peel_tags_to_end()?
                .try_into_commit()?,
        )
    } else {
        stack.base().clone()
    };

    if stack.is_protected(&config) {
        return Err(anyhow!("this branch is protected; rebase is not permitted"));
    }

    stack.check_head_top_mismatch()?;
    let clean_result = stupid.statuses(None)?.check_index_and_worktree_clean();

    let autostash = if matches.get_flag("autostash") {
        true
    } else {
        config
            .plumbing()
            .boolean(
                "branch",
                Some(format!("{branch_name}.stgit").as_str().into()),
                "autostash",
            )
            .transpose()
            .unwrap_or_else(|e| {
                crate::print_warning_message(
                    matches,
                    &format!("Invalid config value `branch.{branch_name}.stgit.autostash`: {e}"),
                );
                Some(false)
            })
            .or_else(|| {
                config
                    .try_boolean("stgit.autostash")
                    .transpose()
                    .unwrap_or_else(|e| {
                        crate::print_warning_message(
                            matches,
                            &format!("Invalid config value `stgit.autostash`: {e}"),
                        );
                        Some(false)
                    })
            })
            .unwrap_or(false)
    };

    let using_stash = if autostash && clean_result.is_err() {
        stupid.stash_push()?;
        true
    } else if let Err(e) = clean_result {
        return Err(e);
    } else {
        false
    };

    let applied = stack.applied().to_vec();

    stack
        .setup_transaction()
        .use_index_and_worktree(true)
        .with_output_stream(get_color_stdout(matches))
        .transact(|trans| {
            trans.pop_patches(|pn| applied.contains(pn))?;
            Ok(())
        })
        .execute("rebase (pop)")?;

    let rebase_cmd = config
        .plumbing()
        .string(
            "branch",
            Some(format!("{branch_name}.stgit").as_str().into()),
            "rebasecmd",
        )
        .or_else(|| config.string("stgit.rebasecmd"))
        .and_then(|bs| bs.to_str().map(str::to_string).ok())
        .unwrap_or_else(|| "git reset --hard".to_string());
    print_info_message(matches, &format!("Rebasing to `{}`", target_commit.id()));
    stupid.user_rebase(&rebase_cmd, target_commit.id)?;

    let stack = Stack::from_branch(&repo, None, InitializationPolicy::RequireInitialized)?;
    let stack = if stack.is_head_top() {
        stack
    } else {
        // Record a new stack state with updated head since the head moved.
        stack.log_external_mods(Some("rebase"))?
    };

    if matches.get_flag("interactive") {
        interactive_pushback(
            stack,
            &repo,
            &config,
            matches,
            &applied,
            allow_push_conflicts,
            committer_date_is_author_date,
        )?;
    } else if !matches.get_flag("nopush") {
        stack.check_head_top_mismatch()?;
        let check_merged = matches.get_flag("merged");
        stack
            .setup_transaction()
            .use_index_and_worktree(true)
            .allow_push_conflicts(allow_push_conflicts)
            .committer_date_is_author_date(committer_date_is_author_date)
            .with_output_stream(get_color_stdout(matches))
            .transact(|trans| trans.push_patches(&applied, check_merged))
            .execute("rebase (reapply)")?;
    }

    if using_stash {
        if stupid.stash_pop()? {
            Ok(())
        } else {
            Err(
                crate::stack::Error::CausedConflicts("stash pop resulted in conflicts".to_string())
                    .into(),
            )
        }
    } else {
        Ok(())
    }
}

const INTERACTIVE_APPLY_LINE: &str = "# --- APPLY_LINE ---";
const INTERACTIVE_HELP_LINES: &str = "\
# Commands:
#
#   k, keep <patch> = do not modify this patch
#   e, edit <patch> = interactively edit this patch
#   s, squash <patch> = squash patch into the previous patch
#   f, fixup <patch> = like \"squash\", but discard this patch's commit message
#   h, hide <patch> = hide patch
#   d, delete <patch> = delete patch
#
# These lines can be reordered; they are executed from top to bottom.
#
# Patches above the APPLY_LINE are applied; other patches are kept unapplied.
";

#[derive(Debug, Clone)]
struct Instruction {
    patchname: PatchName,
    action: Action,
    apply: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Action {
    Keep,
    Edit,
    Squash,
    Fixup,
    Hide,
    Delete,
}

fn interactive_pushback(
    stack: Stack,
    repo: &gix::Repository,
    config: &gix::config::Snapshot,
    matches: &ArgMatches,
    previously_applied: &[PatchName],
    allow_push_conflicts: bool,
    committer_date_is_author_date: bool,
) -> Result<()> {
    let mut stack = stack;

    if stack.all_patches().next().is_none() {
        return Ok(());
    }

    let filename = ".stgit-rebase-interactive.txt";
    std::fs::write(
        filename,
        make_instructions_template(&stack, previously_applied),
    )?;

    let buf = patchedit::call_editor(filename, config)?;
    let buf = buf
        .to_str()
        .map_err(|_| anyhow!("`{filename}` is not valid UTF-8"))?;
    let mut instructions = parse_instructions(buf)?;

    validate_instructions(&stack, &instructions)?;

    let mut index: usize = 0;

    while index < instructions.len() {
        let instruction = instructions[index].clone();
        let patchname = &instruction.patchname;

        match instruction.action {
            Action::Keep => {
                index += 1;
            }

            Action::Delete => {
                // Find contiguous delete instructions in order to delete in batches.
                let mut delete_instructions: Vec<Instruction> = Vec::new();
                while index < instructions.len() && instructions[index].action == Action::Delete {
                    delete_instructions.push(instructions.remove(index));
                }
                assert!(!delete_instructions.is_empty());
                assert!(delete_instructions
                    .iter()
                    .all(|inst| inst.action == Action::Delete));
                let to_delete: Vec<&PatchName> = delete_instructions
                    .iter()
                    .map(|inst| &inst.patchname)
                    .collect();
                stack = stack
                    .setup_transaction()
                    .with_output_stream(get_color_stdout(matches))
                    .transact(|trans| {
                        let popped_extra = trans.delete_patches(|pn| to_delete.contains(&pn))?;
                        assert!(popped_extra.is_empty());
                        Ok(())
                    })
                    .execute("delete")?;
            }

            Action::Hide => {
                let mut hide_instructions: Vec<Instruction> = Vec::new();
                while index < instructions.len() && instructions[index].action == Action::Hide {
                    hide_instructions.push(instructions.remove(index));
                }
                assert!(!hide_instructions.is_empty());
                assert!(hide_instructions
                    .iter()
                    .all(|inst| inst.action == Action::Hide));
                let to_hide: Vec<PatchName> = hide_instructions
                    .iter()
                    .map(|inst| inst.patchname.clone())
                    .collect();
                stack = stack
                    .setup_transaction()
                    .with_output_stream(get_color_stdout(matches))
                    .transact(|trans| trans.hide_patches(&to_hide))
                    .execute("hide")?;
            }

            Action::Edit => {
                let dummy_edit_command = clap::Command::new("dummy-edit");
                let dummy_edit_command = patchedit::add_args(dummy_edit_command, false, false);
                let edit_matches = dummy_edit_command
                    .try_get_matches_from(["dummy-edit", "--edit", "--diff"])
                    .expect("dummy command has valid arguments");
                match patchedit::EditBuilder::default()
                    .original_patchname(Some(patchname))
                    .existing_patch_commit(stack.get_patch_commit(patchname))
                    .allow_diff_edit(true)
                    .edit(&stack, repo, &edit_matches)?
                {
                    patchedit::EditOutcome::TemplateSaved(_) => panic!("template save not enabled"),
                    patchedit::EditOutcome::Edited {
                        new_patchname,
                        new_commit_id,
                    } => {
                        if new_patchname.is_some() || new_commit_id.is_some() {
                            stack = stack
                                .setup_transaction()
                                .committer_date_is_author_date(committer_date_is_author_date)
                                .with_output_stream(get_color_stdout(matches))
                                .transact(|trans| {
                                    let patchname =
                                        if let Some(new_patchname) = new_patchname.as_ref() {
                                            trans.rename_patch(patchname, new_patchname)?;
                                            new_patchname
                                        } else {
                                            patchname
                                        };
                                    if let Some(commit_id) = new_commit_id {
                                        trans.update_patch(patchname, commit_id)?;
                                    }
                                    Ok(())
                                })
                                .execute(&format!("edit: {patchname}"))?;
                        }

                        instructions[index] = Instruction {
                            action: Action::Keep,
                            patchname: new_patchname.unwrap_or_else(|| patchname.clone()),
                            apply: instruction.apply,
                        };

                        index += 1;
                    }
                }
            }

            Action::Squash | Action::Fixup => {
                let action_str = match instruction.action {
                    Action::Squash => "squash",
                    Action::Fixup => "fixup",
                    _ => panic!("only squash and fixup expected"),
                };

                let squash_patchnames: Vec<PatchName> = if index > 0 {
                    let mut patchnames: Vec<PatchName> =
                        vec![instructions[index - 1].patchname.clone()];
                    while index < instructions.len()
                        && instructions[index].action == instruction.action
                    {
                        let Instruction { patchname, .. } = instructions.remove(index);
                        patchnames.push(patchname);
                    }
                    patchnames
                } else {
                    return Err(anyhow!(
                        "cannot {action_str} `{patchname}`: no preceding patch"
                    ));
                };

                let target_patchname = &squash_patchnames[0];

                let dummy_squash_command = clap::Command::new("dummy-squash");
                let dummy_squash_command = patchedit::add_args(dummy_squash_command, true, false);
                let squash_matches = match instruction.action {
                    Action::Squash => {
                        dummy_squash_command.try_get_matches_from(["dummy-squash", "--edit"])
                    }
                    Action::Fixup => {
                        let commit = stack.get_patch_commit(target_patchname);
                        let message = commit.message_raw()?.to_str().map_err(|_| {
                            anyhow!("fixup target patch `{target_patchname}` has non-UTF-8 message")
                        })?;
                        dummy_squash_command.try_get_matches_from([
                            "dummy-squash",
                            "--message",
                            message,
                        ])
                    }
                    _ => panic!("only squash and fixup expected"),
                }
                .expect("dummy command has valid arguments");

                stack = stack
                    .setup_transaction()
                    .with_output_stream(get_color_stdout(matches))
                    .transact(|trans| {
                        let new_patchname = super::squash::squash(
                            trans,
                            &squash_matches,
                            &squash_patchnames,
                            Some(target_patchname),
                            false,
                        )?;
                        instructions[index - 1] = Instruction {
                            action: Action::Keep,
                            patchname: new_patchname,
                            apply: instruction.apply,
                        };
                        Ok(())
                    })
                    .execute("squash")?;
            }
        }
    }

    let to_push: Vec<PatchName> = instructions
        .iter()
        .filter_map(|inst| {
            if inst.apply {
                Some(inst.patchname.clone())
            } else {
                None
            }
        })
        .collect();
    let check_merged = matches.get_flag("merged");

    stack.check_head_top_mismatch()?;
    stack
        .setup_transaction()
        .use_index_and_worktree(true)
        .allow_push_conflicts(allow_push_conflicts)
        .committer_date_is_author_date(committer_date_is_author_date)
        .with_output_stream(get_color_stdout(matches))
        .transact(|trans| trans.push_patches(&to_push, check_merged))
        .execute("rebase (reapply)")?;

    Ok(())
}

fn make_instructions_template(stack: &Stack, previously_applied: &[PatchName]) -> String {
    let name_width = stack.all_patches().map(PatchName::len).max().unwrap();
    let mut template = String::with_capacity(4096);
    let mut found_apply_boundary = false;
    for patchname in stack.all_patches() {
        if !found_apply_boundary && !previously_applied.contains(patchname) {
            writeln!(template, "{INTERACTIVE_APPLY_LINE}").unwrap();
            found_apply_boundary = true;
        }
        let commit = stack.get_patch_commit(patchname);
        let subject = commit
            .message()
            .map(|message_ref| message_ref.title.to_str_lossy())
            .unwrap_or_default();
        writeln!(template, "keep {patchname:name_width$} # {subject}").unwrap();
    }
    if !found_apply_boundary {
        writeln!(template, "{INTERACTIVE_APPLY_LINE}").unwrap();
    }
    template.push_str(INTERACTIVE_HELP_LINES);
    template
}

fn parse_instructions(buf: &str) -> Result<Vec<Instruction>> {
    let mut instructions = Vec::new();
    let mut apply = true;

    for line in buf.lines() {
        let line = line.trim();

        if line.contains(INTERACTIVE_APPLY_LINE) {
            if line == INTERACTIVE_APPLY_LINE {
                apply = false;
            } else {
                return Err(anyhow!("bad APPLY_LINE: `{line}`"));
            }
        }

        let instruction_str = if let Some((instruction_str, _comment)) = line.split_once('#') {
            instruction_str
        } else {
            line
        }
        .trim();

        if instruction_str.is_empty() {
            continue;
        }

        if let Some((action_str, patchname_str)) =
            instruction_str.split_once(|c: char| c.is_ascii_whitespace())
        {
            let action = match action_str {
                "k" | "keep" => Action::Keep,
                "e" | "edit" => Action::Edit,
                "s" | "squash" => Action::Squash,
                "f" | "fix" | "fixup" => Action::Fixup,
                "h" | "hide" => Action::Hide,
                "d" | "delete" => Action::Delete,
                _ => return Err(anyhow!("unknown instruction action `{action_str}`")),
            };

            let patchname = PatchName::from_str(patchname_str)?;

            instructions.push(Instruction {
                patchname,
                action,
                apply,
            });
        } else {
            return Err(anyhow!("bad instruction line: `{line}`"));
        }
    }
    Ok(instructions)
}

fn validate_instructions(stack: &Stack, instructions: &[Instruction]) -> Result<()> {
    let mut seen_patchnames: Vec<&PatchName> = Vec::new();
    for instruction in instructions {
        let patchname = &instruction.patchname;
        if !stack.has_patch(patchname) {
            return Err(anyhow!("unknown patch name `{patchname}`"));
        } else if seen_patchnames.contains(&patchname) {
            return Err(anyhow!("duplicated patch name `{patchname}`"));
        } else {
            seen_patchnames.push(patchname);
        }
    }
    Ok(())
}
